### PR TITLE
cpu/stm32: Fix reception bug in periph_eth

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

The reception code hands RX DMA descriptors back to the DMA right after its contents were copied into the network stack internal buffer. This increases the odds that the DMA never runs out of DMA descriptors to fill, even under high load. However, the loop fetching the Ethernet frame stops to iterate at the end of the frame. If the DMA used one more descriptor to store the FCS, this descriptor was not returned back to the DMA.

### Testing procedure

Send an Ethernet frame that has a size *n* in bytes with *n % 256 > 252*, e.g. 765 bytes. In master, the driver will stop receiving frames after one such frame. E.g.

```
ping -I eth0 -s 703  fe80::XXX
```

produces an Ethernet frame of the correct size for me. With this PR, the issue is fixed.

### Issues/PRs references

Reported by @kfessel via email.

### CODE REVIEW

File: cpu/stm32/periph/eth.c [hunk 1]
```diff
@@ -111,6 +111,53 @@ netdev_t *_netdev;
 static uint8_t _link_state = LINK_STATE_DOWN;
 #endif /* IS_USED(MODULE_STM32_ETH_LINK_UP) */
 
+static void _debug_tx_descriptor_info(void)
+{
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        DEBUG("[stm32_eth] TX descriptors:\n");
```
Comment (miri64) [inline cpu/stm32/periph/eth.c:117]: I know this is highly unusual (though _you_ were the one putting DEBUG commands in a function :stuck_out_tongue_winking_eye:), but as this is now a function and has the potential to be called from multiple places: make calling line an argument to better track where the TX info was that value?

```suggestion
static void _debug_tx_descriptor_info(int line)
{
    if (IS_ACTIVE(ENABLE_DEBUG)) {
        DEBUG("[stm32_eth:%d] TX descriptors:\n", line);
```

File: cpu/stm32/periph/eth.c [hunk 2]
```diff
@@ -480,16 +527,9 @@ static int stm32_eth_send(netdev_t *netdev, const struct iolist *iolist)
 
     /* Error check */
     unsigned i = 0;
+    _debug_tx_descriptor_info();
```
Comment (miri64) [inline cpu/stm32/periph/eth.c:530]:
```suggestion
    _debug_tx_descriptor_info(__LINE__);
```

File: cpu/stm32/periph/eth.c [hunk 3]
```diff
@@ -111,6 +111,53 @@ netdev_t *_netdev;
...
+static inline uint32_t _len_from_rx_desc_status(uint32_t status)
+{
+    /* bits 16-29 contain the frame length including 4 B frame check sequence */
+    return (status >> 16) & 0x3fff;
+}
+
+static void _debug_rx_descriptor_info(void)
+{
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        DEBUG("[stm32_eth] RX descriptors:\n");
```
Comment (miri64) [inline cpu/stm32/periph/eth.c:144]: Here too

```suggestion
static void _debug_rx_descriptor_info(int line)
{
    if (IS_ACTIVE(ENABLE_DEBUG)) {
        DEBUG("[stm32_eth:%d] RX descriptors:\n", line);
```

File: cpu/stm32/periph/eth.c [hunk 4]
```diff
@@ -507,6 +547,7 @@ static int stm32_eth_send(netdev_t *netdev, const struct iolist *iolist)
 
 static int get_rx_frame_size(void)
 {
+    _debug_rx_descriptor_info();
```
Comment (miri64) [inline cpu/stm32/periph/eth.c:550]:
```suggestion
    _debug_rx_descriptor_info(__LINE__);
```

**Review by miri64 — COMMENTED** [review]
Some comments on your DEBUG changes.

**Comment by JannesVolkens** [comment]
I was able to reproduce the described issue in the master and can confirm that this PR fixes it. 

I tested this PR with the nucleo-f207zg.

File: cpu/stm32/periph/eth.c (hunk 1 above)
Comment (maribu) [reply cpu/stm32/periph/eth.c:117]: I like the idea :-) Done

**Review by miri64 — APPROVED** [review]
Code looks good to me, and reasoning sounds sensible. I don't have the hardware to test though, but that was done in https://github.com/RIOT-OS/RIOT/pull/15318#issuecomment-717558514. Please squash

**Comment by kfessel** [comment]
still happening with 

ping -s 706 _IPv6_

or crafted packets of 768 byte fame size

**Comment by maribu** [comment]
> still happening with

can you post the output of `md5sum cpu/stm32/periph/eth.c` in the checked out state you tested with?

**Comment by kfessel** [comment]
RIOT$  md5sum cpu/stm32/periph/eth.c
0213b6694f62380b416e353cc8d05e98  cpu/stm32/periph/eth.c

glaube mir ich hab mehrfach geprüft ob ich im richtigen PR bin und ob ich alles neu gebaut habe

**Comment by maribu** [comment]
I can reproduce the issue with `-s 706`. I will have a second look

**Comment by kfessel** [comment]
Ach

**Comment by maribu** [comment]
The boolean expression was off by one. It should now work as advertised.

**Comment by kfessel** [comment]
I just checked for framesize  768 :heavy_check_mark: 
than checked for framesize 14 (minimal frame 2 x mac(12byte) + type(2 byte)) to framesize 1500 (MTU) for unicast, multicast and broadcast :heavy_check_mark:
parallel ping stayed unimpressed

**Review by miri64 — APPROVED** [review]
Re-ACK, now that @kfessel confirmed the bug is fixed.

**Comment by miri64** [comment]
(please squash)

**Comment by maribu** [comment]
Thanks everyone and especially @kfessel for reporting and testing this fix.

**Comment by maribu** [comment]
@bergzand: Backport provided in https://github.com/RIOT-OS/RIOT/pull/15319

**Pull request merged**